File: 129.cpp
struct Solution {
    int dfs(TreeNode *node, int prefix){
        if (node == nullptr){
            return 0;
        }
        prefix *= 10;
        prefix += node->val;
        if (node->left == nullptr && node->right == nullptr){
            return prefix;
        }
        return dfs(node->left, prefix) + dfs(node->right, prefix);
    }
    inline int sumNumbers(TreeNode* root) {
        return dfs(root, 0);
    }
};

File: sol/129/129.py
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        def traversal(node, val):
            if node is None:
                return 0
            val = val * 10 + node.val
            if node.left is None and node.right is None:
                return val
            return traversal(node.left, val) + traversal(node.right, val)
        return traversal(root, 0)
